### ISSUE TITLE
Applications quick peek and hide if active

### ISSUE THREAD

**Issue by ww7** [zenangst/KeyboardCowboy#547]
I'm fan of [Apptivate](http://www.apptivateapp.com/) keyboard shortcuts manager, it abandoned since 2021 but still works.

Two nice features from Apptivate is missing at KeyboardCowboy:
- quick peek: show window when shortcut hold down (or last in sequence) and hide app window if shortcut is released after some delay of holding
- hide if applications windows already active

**Comment by zenangst** [comment]
Oh, those are both interesting ideas.
Thanks for sharing @ww7, I'll see if these a doable without too much of a rewrite ❤️

**Comment by zenangst** [comment]
Just thought of something that I've used in the past.
So if you have two groups:

- Global
- Specific App (for this example, Let's use Terminal)

The global would for example hold a shortcut to show an application:
`fn-t` -> `Terminal`.

And when Xcode is open, you remap `fn-t` to `Hide`.

You won't get the quick peek, you you'll be able to toggle ("peek") by pushing it once to show and twice to hide.

Hope that helps.

**Comment by zenangst** [comment]
Could you share some more insight to `hide if applications windows already active`?

**Comment by ww7** [comment]
I mean frontmost application, like
Bundle Identifier: org.mozilla.firefoxdeveloperedition
File Path: /Applications/Firefox Developer Edition.app/Contents/MacOS/firefox

**Comment by zenangst** [comment]
So you want to be able to hide the current application?
Could you achieve this with a `Group` that is scoped to that application?
Similar to what I described here? - https://github.com/zenangst/KeyboardCowboy/issues/547#issuecomment-2320159843

**Comment by ww7** [comment]
Yes, thanks! Hiding is not an issue at all, I simply meaning it will be easy to use the same shortcut for hide specific app.
Also `peek` functionality as it implemented it Apptivate app.

**Comment by theprivacydesigner** [comment]
> And when Xcode is open, you remap `fn-t` to `Hide`.

@zenangst Did you mean Terminal (instead of Xcode) here, since the example is around Terminal?

If so, I can’t make your suggestion work.

1. I’ve created a shortcut to open Obsidian with F1
2. I’ve created an Obsidian-specific group and added a new shortcut: hide Obsidian with F1

The first shortcut works, the second one doesn’t.
What am I missing?


<img width="871" alt="Screenshot 2024-10-19 at 5 54 28 PM" src="https://github.com/user-attachments/assets/3cc4bdf0-df97-4764-b4a2-f01e284a565d">

<img width="1122" alt="Screenshot 2024-10-19 at 5 54 43 PM" src="https://github.com/user-attachments/assets/cba4b8ad-be40-4ef4-89f5-38fff00b53f9">

**Comment by zenangst** [comment]
Oh right, yeah Xcode was a typo, meant `Terminal`.

I think you setup looks correct.
I've installed Obsidian and will try and recreate the your desired workflow to check if there is a bug in the mix or if something "non obvious" is missing, because this should work.

What if you try to use the `Hide` command using a keyboard shortcut instead of the application command.

<img width="391" alt="image" src="https://github.com/user-attachments/assets/6b0103ab-e72f-46d7-a194-5c74f0142910">

**Comment by theprivacydesigner** [comment]
That did the trick! Thank you.
For reference:
<img width="1122" alt="Screenshot 2024-10-20 at 10 03 30 AM" src="https://github.com/user-attachments/assets/eef1842e-0f3e-4642-a36e-c53f2df788f0">

**Comment by zenangst** [comment]
@theprivacydesigner Fantastic! I'm still gonna check if there is a bug in the hide application command, so thanks for bringing it to my attention.

**Comment by theprivacydesigner** [comment]
@zenangst just a guess but... in the first setup I tried (the one that didn’t work) I turned on notifications for each command... it looked like _both_ show and hide were triggered when the application was in focus. Which probably is why the app wasn’t hiding.

**Comment by ww7** [comment]
Hide can be implemented regardless to which windows in focus, but via "hidden" status.

**Comment by zenangst** [comment]
Hey @ww7,

Here is some good news for you... I went ahead and implemented application peeking functionality that will be available in the upcoming version.

<img width="1099" alt="image" src="https://github.com/user-attachments/assets/5d9ff032-2584-434a-aefb-ab1ff92897de">

**Comment by ww7** [comment]
Great, thanks! Peek will switch to app if it hold less that 100-400ms?

**Comment by zenangst** [comment]
@ww7 Not currently, the implementation that I made is based on `keyDown` show and then `keyUp` to hide it again when the user releases the key. But I do like the suggestion.

I'm gonna take a stab at it because I can see the potential in it not disappearing if I hold the key down for a longer period of time.

**Comment by zenangst** [comment]
I think I managed to figure it out now.

**Comment by ww7** [comment]
Initial suggestion is wrongly described by me. 

Peek implementation can be in same time as application activation, hiding and peek.
1. Activate after keyDown and keyUp with delay between less than 400ms (or 100-400)
2. Hide if app is frontmost (active) with same behavior as Activate
3. Peek after keyDown and keyUp holded, and hide if keyUp released (but hold longer 400ms)

**Comment by ww7** [comment]
Current default app activation may have "hide if active" flag, this allows to have less shortcuts.

**Comment by zenangst** [comment]
@ww7 3.25.0 now support peeking (sounded way better in my head).

https://github.com/zenangst/KeyboardCowboy/releases/tag/3.25.0

**Comment by ww7** [comment]
@zenangst peeking works nice! 
one request, to add "Hide if active" ("Hide if frontmost") as flag/checkbox or as default behavior for Peek

**Comment by zenangst** [comment]
@ww7 I managed to achieve that behavior by just having a keyboard shortcut scoped to that app.

<img width="1039" alt="Screenshot 2024-10-25 at 10 54 36" src="https://github.com/user-attachments/assets/61172c63-7e7c-43e1-879f-c6c6e2f10f41">
<img width="1039" alt="Screenshot 2024-10-25 at 10 55 01" src="https://github.com/user-attachments/assets/66848603-490a-49bb-8e54-717474bea09d">

I guess that option would save you creating a group and having multiple shortcuts.

**Comment by ww7** [comment]
I have all my apps with shortcuts (~30), it hard to manage Hide for all of them, my previous workflow was:
- hold keys - peek
- press once - show
- press second time - hide
So with press it works also as Peek but without need to hold the keys

As I have two languages (non-latin), all shortcuts doubled...

(offtopic about interface: "About" window have funny `).)` and "Whats new" truncate some peoples) 
:)

**Comment by ww7** [comment]
Initially I thought `peek` and `hide if active/frontmost` was be an flag for Opening

**Comment by zenangst** [comment]
@ww7 Hey @ww7, I had a thought. Instead of making 'Peek' a modifier, let's make it an action. That way, users won't accidentally mix it with 'hide' or 'unhide,' which wouldn't make much sense.

Oh, and 'Peek' also has some special behavior when it comes to 'keyUp' based on a time interval. I didn't want that to be part of the regular path if the user just wants to open an application.

I know this got a bit technical, but I hope you get the idea.

By the way, I think we could still add 'Hide when active/frontmost' as a modifier. I'll check if we can't get that into the next version.

**Comment by ww7** [comment]
Hoping to have Peek with the ability to hide someday :)

**Comment by ww7** [comment]
> By the way, I think we could still add 'Hide when active/frontmost' as a modifier. I'll check if we can't get that into the next version.

Still missing a hide of app via calling same shortcut when it active/frontmost !
